### PR TITLE
[FEATURE] Enable custom actions in V1

### PR DESCRIPTION
**Pull request by cdkini**
Dynamic registry and metaclass will allow users to define their own actions and have them successfully serialize/deserialize without issue.

- [x] Description of PR changes above includes a link to [an existing GitHub issue](https://github.com/great-expectations/great_expectations/issues)
- [x] PR title is prefixed with one of: [BUGFIX], [FEATURE], [DOCS], [MAINTENANCE], [CONTRIB]
- [x] Code is linted - run `invoke lint` (uses `ruff format` + `ruff check`)
- [x] Appropriate tests and docs have been updated

For more information about contributing, visit our [community resources](https://docs.greatexpectations.io/docs/core/introduction/community_resources#contribute-code-or-documentation).

After you submit your PR, keep the page open and **monitor the statuses of the various checks made by our continuous integration process at the bottom of the page. Please fix any issues that come up** and [reach out on Slack](https://greatexpectations.io/slack) if you need help. Thanks for contributing!

### CODE REVIEW

File: great_expectations/checkpoint/actions.py
Comment (cdkini) [inline great_expectations/checkpoint/actions.py:179]: Can this be done cleaner? I want to avoid `ValidationAction` and `DocsAction` base classes that don't have `type`

File: great_expectations/checkpoint/actions.py
Comment (cdkini) [inline great_expectations/checkpoint/actions.py:56]: Do we need custom error types? Ruff says so with our rule we've opted in and it's a low lift.

File: great_expectations/checkpoint/actions.py
Comment (billdirks) [inline great_expectations/checkpoint/actions.py:158]: Since this isn't a constant I would lowercase it.

File: tests/checkpoint/test_checkpoint.py
Comment (billdirks) [inline tests/checkpoint/test_checkpoint.py]: This is what we were doing before but it looks dangerous since it sets some global variable in `set_context(context)` and never undoes it. All our tests should be setting this themselves if needed so maybe this is fine but it feels like we should do this as a context manager and undo it to it's previous state.

File: great_expectations/checkpoint/actions.py
Comment (billdirks) [inline great_expectations/checkpoint/actions.py]: I wonder if we need to specify this this decorator is also on pydantic's `ModelMetaclass`? It's fine if we need it but it would be good if we don't need to be aware of that detail in pydantic.

**Review by billdirks — COMMENTED** [review]
Thanks! This looks good. I left a few comments for consideration.

File: great_expectations/checkpoint/actions.py
Comment (tyler-hoffman) [reply great_expectations/checkpoint/actions.py:56]: I don't feel strongly, but I think it's best practice. Either way, we should probably do what lint says, or otherwise discuss what rules to add/drop external to PRs.

File: great_expectations/checkpoint/actions.py
Comment (tyler-hoffman) [inline great_expectations/checkpoint/actions.py]: > both GX-defined and user-defined ValidationActions
Is this a meaningful distinction?

File: great_expectations/checkpoint/actions.py
Comment (tyler-hoffman) [inline great_expectations/checkpoint/actions.py]: Why pass in the whole config if we only use its `type` field? We should just pass in the `type`. This is really small, but i think it's the correct ™️  thing to do from an API perspective.

File: great_expectations/checkpoint/actions.py
Comment (tyler-hoffman) [reply great_expectations/checkpoint/actions.py:158]: Isn't it though? The object mutable, but the reference is constant.

File: great_expectations/checkpoint/actions.py
Comment (tyler-hoffman) [reply great_expectations/checkpoint/actions.py]: I remember having to do the same thing on expectation metaclasses. I don't recall the details, but I think this is needed.

File: great_expectations/checkpoint/actions.py
Comment (tyler-hoffman) [inline great_expectations/checkpoint/actions.py:122]: it'd be nice if `type` were a `ClassVar` on the action class and we could just pass in the class. Looks like that's not the case though?

File: tests/checkpoint/test_checkpoint.py
Comment (tyler-hoffman) [reply tests/checkpoint/test_checkpoint.py]: Making it a generator fixture would mean you could do both the setup teardown, plus you wouldn't have to explicitly pass in `mocker`, which is a slight ergonomics benefit.

File: tests/checkpoint/test_checkpoint.py
Comment (tyler-hoffman) [inline tests/checkpoint/test_checkpoint.py]: I think we know this to be the case, but an isinstance check probably gives us a slight bit more confidence here.

**Review by tyler-hoffman — COMMENTED** [review]
Looks great to me! I'd love to see the slight tweak to `get_from_config`'s signature, but that's not blocking. The only reason I'm not approving is because AFAIK this will 422 against cloud, so we need to get the backend changes in first. LMK if I'm mistaken!

**Review by tyler-hoffman — CHANGES_REQUESTED** [review]
Actually I'm going to throw a blocker on here just to make sure we don't accidentally release something that breaks against cloud. I'll approve once cloud is ready for this!

File: great_expectations/checkpoint/actions.py
Comment (cdkini) [reply great_expectations/checkpoint/actions.py]: It's what enables extensibility so I thought to call it out - happy to remove the distinction if we don't see it adding value.

Comment (tyler-hoffman) [reply great_expectations/checkpoint/actions.py]: My thought is that it makes an arbitrary distinction, which could cause confusion. Your call on if you agree or not. Or maybe get another opinion. IDK, it's totally non-blocking

File: great_expectations/checkpoint/actions.py
Comment (cdkini) [reply great_expectations/checkpoint/actions.py:122]: Ah no and I think that changes some serialization behavior? I'd vote to not update that here

File: great_expectations/checkpoint/actions.py
Comment (cdkini) [reply great_expectations/checkpoint/actions.py]: I was thinking the internals around `type` should be encapsulated within the action file and related classes but I agree, this feels a little inelegant.

File: great_expectations/checkpoint/actions.py
Comment (cdkini) [reply great_expectations/checkpoint/actions.py:158]: I could go either way but my vote is to leave it as is

File: great_expectations/checkpoint/actions.py
Comment (cdkini) [reply great_expectations/checkpoint/actions.py]: Enables type hinting (breaks otherwise) - added a comment

File: tests/checkpoint/test_checkpoint.py
Comment (cdkini) [reply tests/checkpoint/test_checkpoint.py]: Made a fixture

File: great_expectations/checkpoint/actions.py
Comment (billdirks) [reply great_expectations/checkpoint/actions.py:158]: When I say a constant object I mean an immutable object; eg what you would get in a data class when you set `frozen=True`. I think that makes it easier to reason about a variable thats all caps means.

File: great_expectations/checkpoint/actions.py
Comment (billdirks) [reply great_expectations/checkpoint/actions.py]: I agree. I think it's clearer to write "Registers actions" or something. Then a reader doesn't even need to think if there is a difference between the two (and there isn't).

File: great_expectations/checkpoint/actions.py
Comment (tyler-hoffman) [inline great_expectations/checkpoint/actions.py:138]: Doesn't this syntax not work for 3.9?

Comment (cdkini) [reply great_expectations/checkpoint/actions.py:138]: `str | None` vs `Optional[str]`? I think that's fine?

**Review by tyler-hoffman — APPROVED** [review]
Looks great!

File: great_expectations/checkpoint/actions.py
Comment (tyler-hoffman) [reply great_expectations/checkpoint/actions.py:138]: I guess you're right! I have no idea what I'm doing!

**Pull request merged**
